Test matrix in CI

noahsa opened the pull request:
Added Github Action to automate tests on different Python versions/ platforms:

1) Ubuntu Latest - 3.6, 3.7, 3.8
2) Mac OS - 3.6, 3.7

Partially resolves #4.

Here are some additional features not in the PR. I would love to discuss which features should be added prior to PR being merged. Or if they are better addressed in future PRs. They're all not as trivial as adding them to the matrix specification. Caching may be the most copy/paste like. 1 & 3 may require additional investigation in to dependency specification.  

1. Expand matrix to Windows
2. Caching of Python dependencies.
3. Python 3.5 compatibility

carlomazzaferro commented:
🚀🚀🚀

I believe starting small (i.e. what you have now -- ubuntu 3.6/7/8 and MacOS 3.7/8) to be a good start. Long-term, I think having tests for the latest 3 python versions to be a good compromise, so I wouldn't bother with 3.5 (which is great as it has less support for type hints).

We can think of adding Windows support as well, but that's a bigger push that's likely best left to its own PR.

What issue do you envision with caching the linting deps? That works as its own step, so it shouldn't require any testing matrix support.

noahsa commented on .github/workflows/workflow.yaml:
```diff
@@ -39,25 +39,38 @@ jobs:
-
-      - name: Run unit tests
+          path: ${{ matrix.path }}
+          key: ${{ runner.os }}-pip-${{ hashFiles('reqs/base.txt') }}
```
@carlomazzaferro One thing to note in the caching process for the testing matrix is the hash on 'reqs/base.txt'. I think it'll only establish a new cache when reqs/base.txt changes. So if any of the dependencies in reqs/extras change it may not install the new dependencies. What are your thoughts on this? Also, want to check that I am thinking about this correctly :)

carlomazzaferro replied:
The overall approach looks sound to me (I'm still fairly new to GHA caching) -- perhaps `${{ hashFiles('setup.py','reqs/all.txt') }}` will provide the behaviour we want. It will update the cache when any dep is changed, or when `setup.py` changes (when bumping a new scikit-hts version). The latter is not strictly necessary, but probably a good practice. Wdyt?

noahsa replied:
I like it, it makes sense to me. Making change

noahsa commented:
Awesome, I like that approach!

+1 to windows being in its own PR.

I don't see any issues with caching the linter dependencies. My perceived complexity probably comes from being new to using  it. 

I added back the lint cache that you already had in place. I just added a cache to the dependency installs for the test matrix, to speed up this process too. Even though waiting a couple minutes for CI isn't a huge deal.

Tagged you in a comment, for one thing I was questioning in the setup for the testing matrix cache.

carlomazzaferro commented:
> Awesome, I like that approach!
> 
> +1 to windows being in its own PR.
> 
> I don't see any issues with caching the linter dependencies. My perceived complexity probably comes from being new to using it.
> 
> I added back the lint cache that you already had in place. I just added a cache to the dependency installs for the test matrix, to speed up this process too. Even though waiting a couple minutes for CI isn't a huge deal.
> 
> Tagged you in a comment, for one thing I was questioning in the setup for the testing matrix cache. If we're too unsure about this, we can remove the caching from this piece and handle it later.

We can adopt the caching you implemented, or add in my suggestion. We can optimise later if needed, but as you point out this isn't a huge deal for now. 

Overall this looks good, waiting on your reply on the comment and then we can merge

noahsa commented:
Added in your suggestion. I think it solves my concern. Like you said we can always optimize later.

CI passing. Not planning to make any more changes. Will close #4 when merged.

Thanks!!

carlomazzaferro reviewed (approved):
LGTM
